sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+                            .ConfigureAwait(false);
+
+                        InstanceOwnership = null;
+                        ActivePartitionProcessors = null;
+                        ActivePartitionProcessorTokenSources = null;
+
+                        // TODO: once IsRunning is implemented, update the following comment.
+                        // We need to wait until all tasks have stopped before making the load balancing task null.  If we did it sooner, we
+                        // would have a race condition where the user could update the processing handlers while some pumps are still running.
+
+                        ActiveLoadBalancingTask = null;
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Performs load balancing between multiple <see cref="EventProcessorClient" /> instances, claiming others' partitions to enforce
+        ///   a more equal distribution when necessary.  It also manages its own partition processing tasks and ownership.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task RunAsync(CancellationToken cancellationToken)
+        {
+            // We'll use this connection to retrieve an updated list of partition ids from the service.
+
+            await using var connection = CreateConnection();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                Stopwatch cycleDuration = Stopwatch.StartNew();
+
+                // Renew this instance's ownership so they don't expire.
+                // TODO: renew only after retrieving updated ownership so we use updated checkpoints.
+
+                await RenewOwnershipAsync().ConfigureAwait(false);
+
+                // From the storage service provided by the user, obtain a complete list of ownership, including expired ones.  We may still need
+                // their eTags to claim orphan partitions.
+
+                var completeOwnershipList = (await ListOwnershipAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup)
+                    .ConfigureAwait(false))
+                    .ToList();
+
+                // Filter the complete ownership list to obtain only the ones that are still active.  The expiration time defaults to 30 seconds,
+                // but it may be overridden by a derived class.
+
+                IEnumerable<PartitionOwnership> activeOwnership = completeOwnershipList
+                    .Where(ownership => DateTimeOffset.UtcNow.Subtract(ownership.LastModifiedTime.Value) < OwnershipExpiration);
```

Comment:
You probably want to capture `Now` outside of the calculation so that the same value is applied to all checks when filtering.